src/test/test_switches.cpp

```diff
@@ -0,0 +1,315 @@
+        TEST_ASSERT_EQUAL(expected, Radio.TXdataBuffer[i+1]);
+    }
+
+    // byte 5 contains some bits from the first three channels
+    expected = (crsf.ChannelDataIn[0] & 0b111) << 5 | (crsf.ChannelDataIn[1] & 0b111) << 2 | (crsf.ChannelDataIn[2] & 0b110) >> 1;
+    TEST_ASSERT_EQUAL(expected, Radio.TXdataBuffer[5]);
+
+    // bit 7 of byte 6 contains the lsb of ChannelDataIn[2], 
+    expected = crsf.ChannelDataIn[2] & 0b1;
+    TEST_ASSERT_EQUAL(expected, (Radio.TXdataBuffer[6] & 0b10000000) >> 7);
+
+    // bits 6,5 contain bits 1,2 of ChannelDataIn[3]
+    expected = (crsf.ChannelDataIn[3] & 0b110) >> 1;
+    TEST_ASSERT_EQUAL(expected, (Radio.TXdataBuffer[6] & 0b1100000) >> 5);
+
+    // the sequential switch bits:
+    // expect index in 2-4 and value in 0,1
+    TEST_ASSERT_EQUAL(3, (Radio.TXdataBuffer[6] & 0b11100)>>2);
+    TEST_ASSERT_EQUAL(crsf.currentSwitches[3], Radio.TXdataBuffer[6] & 0b11);
+}
+
+// decoding
+/* Check the decoding of a packet after rx
+*/
+void test_decodingSEQ()
+{
+    uint8_t UID[6] = {MY_UID};
+    uint8_t DeviceAddr = UID[5] & 0b111111;
+    // uint8_t expected;
+
+    // Define the input data
+    // 4 channels of analog data
+    crsf.ChannelDataIn[0] = 0x0123;
+    crsf.ChannelDataIn[1] = 0x4567;
+    crsf.ChannelDataIn[2] = 0x89AB;
+    crsf.ChannelDataIn[3] = 0xCDEF;
+
+    // 8 switches
+    for(int i=0; i<N_SWITCHES; i++) {
+        crsf.currentSwitches[i] =  (i+1) % 3;
+        crsf.sentSwitches[i] = (i+1) % 3; // make all the sent values match
+    }
+
+    // set the nextSwitchIndex so we know which switch to expect in the packet
+    crsf.nextSwitchIndex=3;
+
+    // use the encoding method to pack it into Radio.TXdataBuffer
+    GenerateChannelDataSeqSwitch(&Radio, &crsf, DeviceAddr);
+
+    // copy into the required buffer for the decoder
+    memcpy((void*)Radio.RXdataBuffer, (const void*)Radio.TXdataBuffer, sizeof(Radio.RXdataBuffer));
+
+    // clear the output buffer to avoid cross-talk between tests
+    memset((void *) &crsf.PackedRCdataOut, 0, sizeof(crsf.PackedRCdataOut));
+
+    // run the decoder, results in crsf->PackedRCdataOut
+    UnpackChannelDataSeqSwitches(&Radio, &crsf);
+
+    // compare the unpacked results with the input data
+    TEST_ASSERT_EQUAL(crsf.ChannelDataIn[0] & 0b11111111111, crsf.PackedRCdataOut.ch0); // contains 11 bit data
+    TEST_ASSERT_EQUAL(crsf.ChannelDataIn[1] & 0b11111111111, crsf.PackedRCdataOut.ch1); // contains 11 bit data
+    TEST_ASSERT_EQUAL(crsf.ChannelDataIn[2] & 0b11111111111, crsf.PackedRCdataOut.ch2); // contains 11 bit data
+    TEST_ASSERT_EQUAL(crsf.ChannelDataIn[3] & 0b11111111110, crsf.PackedRCdataOut.ch3); // last analog channel is truncated to 10 bits
+
+    TEST_ASSERT_EQUAL(SWITCH2b_to_CRSF(crsf.currentSwitches[3] & 0b11), crsf.PackedRCdataOut.ch7); // We forced switch 3 to be sent as the sequential field
+}
+
+
+#endif // TARGET_TTGO_LORA_V1_AS_TX
+
+void setup_switches()
```

Comment:
Very nice work on the tests for this. LOTS of good coverage including the packet encoding / decoding stuff.

One thing I'd like us to try and adopt for the tests is using comments in the gherkin test syntax:
https://cucumber.io/docs/gherkin/reference/

This is the syntax that BF use in their tests, e.g see:
https://github.com/betaflight/betaflight/blob/a2aa3f19f44855451153494f87eba4b3098d01a3/src/test/unit/osd_unittest.cc#L145

Reply:
This looks interesting - I haven't seen it before.